File: app/src/main/java/com/zibete/proyecto1/utils/UserRepository.java
package com.zibete.proyecto1.utils;

import static com.zibete.proyecto1.Constants.Empty;
import static com.zibete.proyecto1.Constants.chatWith;
import static com.zibete.proyecto1.utils.FirebaseRefs.auth;
import static com.zibete.proyecto1.utils.FirebaseRefs.ref_cuentas;
import static com.zibete.proyecto1.utils.FirebaseRefs.ref_datos;
import static com.zibete.proyecto1.utils.FirebaseRefs.user;

import android.annotation.SuppressLint;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.graphics.Typeface;
import android.location.Location;
import android.view.ContextThemeWrapper;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;

import com.google.android.material.snackbar.Snackbar;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.ValueEventListener;
import com.zibete.proyecto1.model.ChatWith;
import com.zibete.proyecto1.model.State;
import com.zibete.proyecto1.R;

import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Objects;

// Clase que maneja la comunicación con Firebase para datos de usuario
// (repository = repositorio → capa de acceso a datos)
public class UserRepository {


    // ------------------------------------------------------------
    // setUserOnline → establece al usuario como conectado
    // (StateOnLine = poner en línea)
    // ------------------------------------------------------------
    public static void setUserOnline(Context context, String id_user) {

        // 🔸 Crear el objeto Estado con los textos de recursos
        // Estado = clase de modelo que representa el estado de conexión
        State currentState = new State(
                context.getString(R.string.conectado), // "connected" → conectado
                "",
                ""
        );

        // 🔸 Guardar el estado en Firebase (dos ubicaciones distintas)
        ref_datos.child(id_user).child("Estado").setValue(currentState);
        ref_cuentas.child(id_user).child("estado").setValue(true);
    }

    public static void setUserOffline(Context context, String id_user){

        if (user != null) {
            final Calendar c = Calendar.getInstance();
            final SimpleDateFormat timeFormat = new SimpleDateFormat("HH:mm");
            final SimpleDateFormat dateFormat = new SimpleDateFormat("dd/MM/yyyy");

            final State cState = new State(
                    context.getString(R.string.ultVez),
                    dateFormat.format(c.getTime()),
                    timeFormat.format(c.getTime()));

            ref_datos.child(id_user).child("Estado").setValue(cState);
            ref_cuentas.child(id_user).child("estado").setValue(false);
        }

    }

    public static void stateUser(
            final Context context,
            final String userId,                    // id_user → usuario a mostrar
            final ImageView iconConnected,          // icon_conectado
            final ImageView iconDisconnected,       // icon_desconectado
            final TextView tvStatus,                // tv_estado
            final String type                       // type → ej. "chatWith"
    ) {
        ref_datos.child(userId).child("Estado").addValueEventListener(new ValueEventListener() {
            @SuppressLint("SetTextI18n")
            @Override
            public void onDataChange(@NonNull final DataSnapshot dataSnapshot) {

                if (dataSnapshot.exists()) {

                    Calendar c = Calendar.getInstance();
                    SimpleDateFormat dateFormat = new SimpleDateFormat("dd/MM/yyyy");

                    final String estado = dataSnapshot.child("estado").getValue(String.class);   // estado → "conectado", "escribiendo", etc.
                    final String fecha = dataSnapshot.child("fecha").getValue(String.class);     // fecha última vez
                    final String hora = dataSnapshot.child("hora").getValue(String.class);       // hora última vez

                    if (estado != null && estado.equals(context.getString(R.string.conectado))) {
                        // ✅ Conectado (online)
                        iconConnected.setVisibility(View.VISIBLE);
                        iconDisconnected.setVisibility(View.GONE);
                        tvStatus.setText(context.getString(R.string.enlinea)); // "en línea"
                        tvStatus.setTypeface(null, Typeface.NORMAL);
                        tvStatus.setTextColor(context.getResources().getColor(R.color.colorClaro));

                    } else {
                        // ⚠️ Puede estar escribiendo / grabando / desconectado

                        if (estado != null &&
                                (estado.equals(context.getString(R.string.escribiendo)) ||
                                        estado.equals(context.getString(R.string.grabando)))) {

                            // 📝 Escribiendo / 🎙 Grabando
                            // Tenemos que chequear si el chat actual es conmigo
                            ref_datos.child(userId)
                                    .child("ChatList")
                                    .child("Actual")
                                    .addValueEventListener(new ValueEventListener() {
                                        @Override
                                        public void onDataChange(@NonNull DataSnapshot dsChat) {
                                            if (dsChat.exists()) {

                                                String currentChat = dsChat.getValue(String.class);
                                                // usuario actual logueado
                                                String myUid = auth.getCurrentUser() != null
                                                        ? auth.getCurrentUser().getUid()
                                                        : null;

                                                // compara: if (Actual == userLogged + type)
                                                if (myUid != null && currentChat != null
                                                        && currentChat.equals(myUid + type)) {

                                                    // ✅ Mostrar “escribiendo” o “grabando”
                                                    iconConnected.setVisibility(View.VISIBLE);
                                                    iconDisconnected.setVisibility(View.GONE);
                                                    tvStatus.setText(estado);
                                                    tvStatus.setTypeface(null, Typeface.ITALIC);
                                                    tvStatus.setTextColor(context.getResources().getColor(R.color.accent));

                                                } else {
                                                    // está escribiendo pero NO conmigo → mostrar online normal
                                                    iconConnected.setVisibility(View.VISIBLE);
                                                    iconDisconnected.setVisibility(View.GONE);
                                                    tvStatus.setText(context.getString(R.string.enlinea));
                                                    tvStatus.setTypeface(null, Typeface.NORMAL);
                                                    tvStatus.setTextColor(context.getResources().getColor(R.color.colorClaro));
                                                }
                                            }
                                        }

                                        @Override
                                        public void onCancelled(@NonNull DatabaseError error) {
                                            // opcional: log
                                        }
                                    });

                        } else {
                            // ⏱ Última vez conectado
                            iconConnected.setVisibility(View.GONE);
                            iconDisconnected.setVisibility(View.VISIBLE);
                            tvStatus.setTypeface(null, Typeface.NORMAL);
                            tvStatus.setTextColor(context.getResources().getColor(R.color.colorClaro));

                            if (fecha != null && fecha.equals(dateFormat.format(c.getTime()))) {
                                // hoy
                                tvStatus.setText(
                                        context.getString(R.string.ultVez) + " " +
                                                context.getString(R.string.hoy) + " " +
                                                context.getString(R.string.a_las) + " " +
                                                hora
                                );
                            } else {
                                // quizá ayer
                                Calendar calendar = Calendar.getInstance();
                                calendar.add(Calendar.DATE, -1);

                                if (fecha != null && fecha.equals(dateFormat.format(calendar.getTime()))) {
                                    tvStatus.setText(
                                            context.getString(R.string.ultVez) + " " +
                                                    context.getString(R.string.ayer) + " " +
                                                    context.getString(R.string.a_las) + " " +
                                                    hora
                                    );
                                } else {
                                    // fecha cualquiera
                                    tvStatus.setText(
                                            context.getString(R.string.ultVez) + " " +
                                                    fecha + " " +
                                                    context.getString(R.string.a_las) + " " +
                                                    hora
                                    );
                                }
                            }
                        }
                    }
                } else {
                    // no hay nodo "Estado" → mostrar desconectado
                    iconConnected.setVisibility(View.GONE);
                    iconDisconnected.setVisibility(View.VISIBLE);
                    tvStatus.setText(context.getString(R.string.desconectado)); // "disconnected" → desconectado
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {
                // opcional: log
            }
        });
    }

    public static void setNoLeido(String id_user, String type){

        ref_datos.child(user.getUid()).child(type).child(id_user).child("noVisto").addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull final DataSnapshot dataVistos) {
                if (dataVistos.exists()) {
                    final Integer noVistos = dataVistos.getValue(Integer.class);

                    ref_datos.child(user.getUid()).child("ChatList").child("msgNoLeidos").addListenerForSingleValueEvent(new ValueEventListener() {
                        @Override
                        public void onDataChange(@NonNull DataSnapshot dataLeidos) {
                            if (dataLeidos.exists()) {
                                final Integer noLeidos = dataLeidos.getValue(Integer.class);
                                Integer count = noLeidos - noVistos;
                                if ((noVistos > 0)) {
                                    dataVistos.getRef().setValue(0);
                                    dataLeidos.getRef().setValue(count);
                                } else {
                                    dataVistos.getRef().setValue(1);
                                    dataLeidos.getRef().setValue(noLeidos + 1);
                                }
                            }
                        }
                        @Override
                        public void onCancelled(@NonNull DatabaseError error) {
                        }
                    });
                }
            }
            @Override
            public void onCancelled(@NonNull DatabaseError error) {
            }
        });
    }

    public static void Silent(final String name_user, final String id_user, final String type){


        ref_datos.child(user.getUid()).child(type).child(id_user).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {

                if (dataSnapshot.exists()) {

                    String state = dataSnapshot.child("estado").getValue(String.class);
                    String photo = dataSnapshot.child("wUserPhoto").getValue(String.class);

                    if (photo.equals(Empty)){

                        dataSnapshot.getRef().removeValue();

                    }else{

                        if (state.equals("silent")) {

                            dataSnapshot.getRef().child("estado").setValue(type);

                        }else{

                            dataSnapshot.getRef().child("estado").setValue("silent");

                        }
                    }

                }else {

                    newChatWith(dataSnapshot, id_user, name_user, "silent");

                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) {
            }
        });


    }

    public static void newChatWith(@NonNull DataSnapshot dataSnapshot, String id_user, String name_user, String state) {
        final Calendar c = Calendar.getInstance();
        final SimpleDateFormat dateFormat3 = new SimpleDateFormat("dd/MM/yyyy HH:mm:ss:SS");
        final ChatWith newChat = new ChatWith(
                "",
                dateFormat3.format(c.getTime()),
                null,
                "",
                id_user,
                name_user,
                Empty,
                state,
                0,
                1);

        dataSnapshot.getRef().setValue(newChat);
    }

    public static void setBlockUser(
            final Context context,
            final String name_user,
            final String id_user,
            final View view,
            final String type)
    {

        AlertDialog.Builder builder = new AlertDialog.Builder(new ContextThemeWrapper(context, R.style.AlertDialogApp));
        builder.setTitle("Bloquear");
        builder.setMessage("¿Desea bloquear a " + name_user + "?");

        builder.setCancelable(false);
        builder.setPositiveButton("Aceptar", new DialogInterface.OnClickListener() {

            public void onClick(DialogInterface builder, int id) {

                ref_datos.child(user.getUid()).child(type).child(id_user).addListenerForSingleValueEvent(new ValueEventListener() {
                    @Override
                    public void onDataChange(@NonNull final DataSnapshot dataSnapshot) {

                        if (dataSnapshot.exists()){

                            dataSnapshot.getRef().child("estado").setValue("bloq");

                        } else {

                            newChatWith(dataSnapshot, id_user, name_user, "bloq");

                        }
                    }
                    @Override
                    public void onCancelled(@NonNull DatabaseError error) {
                    }
                });

                final Snackbar snack = Snackbar.make(view, "Bloqueaste a " + name_user +", podrás desbloquearlo cuando desees", Snackbar.LENGTH_INDEFINITE);
                snack.setAction("OK", new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        snack.dismiss();
                    }
                });
                snack.setBackgroundTint(context.getResources().getColor(R.color.colorC));
                TextView tv = snack.getView().findViewById(com.google.android.material.R.id.snackbar_text);
                tv.setTextAlignment(View.TEXT_ALIGNMENT_CENTER);
                snack.show();

            }
        });
        builder.setNegativeButton("Cancelar", new DialogInterface.OnClickListener() {
            public void onClick(DialogInterface builder, int id) {
                return;
            }
        });
        builder.show();

    }

    public static void setUnBlockUser (
            final Context context,
            final String id_user,
            final String name_user,
            final View view,
            final String type) {

        new AlertDialog.Builder(new ContextThemeWrapper(context, R.style.AlertDialogApp))

                .setMessage("¿Desea desbloquear a " + name_user + "?")
                .setPositiveButton("Aceptar", new DialogInterface.OnClickListener() {
                    public void onClick(DialogInterface builder, int id) {

                        ref_datos.child(user.getUid()).child(type).child(id_user).addListenerForSingleValueEvent(new ValueEventListener() {
                            @Override
                            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {

                                String photo = dataSnapshot.child("wUserPhoto").getValue(String.class);
                                if (photo.equals(Empty)){
                                    dataSnapshot.getRef().removeValue();
                                }else{
                                    dataSnapshot.getRef().child("estado").setValue(type);

                                }
                            }

                            @Override
                            public void onCancelled(@NonNull DatabaseError error) {

                            }
                        });

                        Snackbar snack = Snackbar.make(view, "Desbloqueaste a " + name_user, Snackbar.LENGTH_SHORT);
                        snack.setBackgroundTint(context.getResources().getColor(R.color.colorC));
                        TextView tv = snack.getView().findViewById(com.google.android.material.R.id.snackbar_text);
                        tv.setTextAlignment(View.TEXT_ALIGNMENT_CENTER);
                        snack.show();

                    }
                })
                .setNegativeButton("Cancelar", new DialogInterface.OnClickListener() {

                    public void onClick(DialogInterface builder, int id) {
                        return;
                    }
                })
                .show();

    }




    public static void bindBlockStatus (
            String user_id,
            final ImageView profile_bloc)

    { // bindBlockStatus = vincular estado de bloqueo

        ref_datos.child(user.getUid()).child(chatWith).child(user_id).child("estado").addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {

                if (Objects.equals(dataSnapshot.getValue(String.class), "bloq")) {
                    profile_bloc.setVisibility(View.VISIBLE);
                } else {
                    profile_bloc.setVisibility(View.GONE);
                }
            }
            @Override
            public void onCancelled(@NonNull DatabaseError error) {
            }
        });
    }

    public static Double latitude;
    public static Double longitude;
    public static void updateLocationUI(Location mLastLocation) {
        if (mLastLocation == null) return;

        latitude = mLastLocation.getLatitude();
        longitude = mLastLocation.getLongitude();

        ref_cuentas.child(user.getUid()).addListenerForSingleValueEvent(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull DataSnapshot dataSnapshot) {
                if (dataSnapshot.exists()) {
                    FirebaseRefs.ref_cuentas.child(user.getUid()).child("latitud").setValue(latitude);
                    FirebaseRefs.ref_cuentas.child(user.getUid()).child("longitud").setValue(longitude);
                }
            }

            @Override
            public void onCancelled(@NonNull DatabaseError error) { }
        });
    }





}

File: app/src/main/java/com/zibete/proyecto1/utils/UserRepository.kt
package com.zibete.proyecto1.utils

import android.annotation.SuppressLint
import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import android.graphics.Typeface
import android.location.Location
import android.view.ContextThemeWrapper
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import com.google.android.material.snackbar.Snackbar
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import com.google.firebase.database.ValueEventListener
import com.zibete.proyecto1.R
import com.zibete.proyecto1.model.ChatWith
import com.zibete.proyecto1.model.State
import com.zibete.proyecto1.utils.FirebaseRefs.auth
import com.zibete.proyecto1.utils.FirebaseRefs.refCuentas
import com.zibete.proyecto1.utils.FirebaseRefs.refDatos
import com.zibete.proyecto1.utils.FirebaseRefs.user
import java.text.SimpleDateFormat
import java.util.Calendar

// Clase que maneja la comunicación con Firebase para datos de usuario
// (repository = repositorio → capa de acceso a datos)
object UserRepository {
    // ------------------------------------------------------------
    // setUserOnline → establece al usuario como conectado
    // (StateOnLine = poner en línea)
    // ------------------------------------------------------------
    @JvmStatic
    fun setUserOnline(context: Context, id_user: String) {
        // 🔸 Crear el objeto Estado con los textos de recursos
        // Estado = clase de modelo que representa el estado de conexión

        val currentState = State(
            context.getString(R.string.conectado),  // "connected" → conectado
            "",
            ""
        )

        // 🔸 Guardar el estado en Firebase (dos ubicaciones distintas)
        refDatos.child(id_user).child("Estado").setValue(currentState)
        refCuentas.child(id_user).child("estado").setValue(true)
    }

    @JvmStatic
    fun setUserOffline(context: Context, id_user: String) {
        if (user != null) {
            val c = Calendar.getInstance()
            val timeFormat = SimpleDateFormat("HH:mm")
            val dateFormat = SimpleDateFormat("dd/MM/yyyy")

            val cState = State(
                context.getString(R.string.ultVez),
                dateFormat.format(c.getTime()),
                timeFormat.format(c.getTime())
            )

            refDatos.child(id_user).child("Estado").setValue(cState)
            refCuentas.child(id_user).child("estado").setValue(false)
        }
    }

    @JvmStatic
    fun stateUser(
        context: Context,
        userId: String,  // id_user → usuario a mostrar
        iconConnected: ImageView,  // icon_conectado
        iconDisconnected: ImageView,  // icon_desconectado
        tvStatus: TextView,  // tv_estado
        type: String? // type → ej. "chatWith"
    ) {
        refDatos.child(userId).child("Estado").addValueEventListener(object : ValueEventListener {
            @SuppressLint("SetTextI18n")
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                if (dataSnapshot.exists()) {
                    val c = Calendar.getInstance()
                    val dateFormat = SimpleDateFormat("dd/MM/yyyy")

                    val estado = dataSnapshot.child("estado")
                        .getValue<String?>(String::class.java) // estado → "conectado", "escribiendo", etc.
                    val fecha = dataSnapshot.child("fecha")
                        .getValue<String?>(String::class.java) // fecha última vez
                    val hora = dataSnapshot.child("hora")
                        .getValue<String?>(String::class.java) // hora última vez

                    if (estado != null && estado == context.getString(R.string.conectado)) {
                        // ✅ Conectado (online)
                        iconConnected.setVisibility(View.VISIBLE)
                        iconDisconnected.setVisibility(View.GONE)
                        tvStatus.setText(context.getString(R.string.enlinea)) // "en línea"
                        tvStatus.setTypeface(null, Typeface.NORMAL)
                        tvStatus.setTextColor(context.getResources().getColor(R.color.colorClaro))
                    } else {
                        // ⚠️ Puede estar escribiendo / grabando / desconectado

                        if (estado != null &&
                            (estado == context.getString(R.string.escribiendo) ||
                                    estado == context.getString(R.string.grabando))
                        ) {
                            // 📝 Escribiendo / 🎙 Grabando
                            // Tenemos que chequear si el chat actual es conmigo

                            refDatos.child(userId)
                                .child("ChatList")
                                .child("Actual")
                                .addValueEventListener(object : ValueEventListener {
                                    override fun onDataChange(dsChat: DataSnapshot) {
                                        if (dsChat.exists()) {
                                            val currentChat =
                                                dsChat.getValue<String?>(String::class.java)
                                            // usuario actual logueado
                                            val myUid: String? = if (auth.currentUser != null)
                                                auth.currentUser?.uid
                                            else
                                                null

                                            // compara: if (Actual == userLogged + type)
                                            if (myUid != null && currentChat != null && currentChat == myUid + type) {
                                                // ✅ Mostrar “escribiendo” o “grabando”

                                                iconConnected.visibility = View.VISIBLE
                                                iconDisconnected.visibility = View.GONE
                                                tvStatus.text = estado
                                                tvStatus.setTypeface(null, Typeface.ITALIC)
                                                tvStatus.setTextColor(
                                                    context.resources.getColor(R.color.accent)
                                                )
                                            } else {
                                                // está escribiendo pero NO conmigo → mostrar online normal
                                                iconConnected.visibility = View.VISIBLE
                                                iconDisconnected.visibility = View.GONE
                                                tvStatus.text = context.getString(R.string.enlinea)
                                                tvStatus.setTypeface(null, Typeface.NORMAL)
                                                tvStatus.setTextColor(
                                                    context.resources
                                                        .getColor(R.color.colorClaro)
                                                )
                                            }
                                        }
                                    }

                                    override fun onCancelled(error: DatabaseError) {
                                        // opcional: log
                                    }
                                })
                        } else {
                            // ⏱ Última vez conectado
                            iconConnected.setVisibility(View.GONE)
                            iconDisconnected.setVisibility(View.VISIBLE)
                            tvStatus.setTypeface(null, Typeface.NORMAL)
                            tvStatus.setTextColor(
                                context.getResources().getColor(R.color.colorClaro)
                            )

                            if (fecha != null && fecha == dateFormat.format(c.getTime())) {
                                // hoy
                                tvStatus.setText(
                                    context.getString(R.string.ultVez) + " " +
                                            context.getString(R.string.today) + " " +
                                            context.getString(R.string.a_las) + " " +
                                            hora
                                )
                            } else {
                                // quizá ayer
                                val calendar = Calendar.getInstance()
                                calendar.add(Calendar.DATE, -1)

                                if (fecha != null && fecha == dateFormat.format(calendar.getTime())) {
                                    tvStatus.setText(
                                        context.getString(R.string.ultVez) + " " +
                                                context.getString(R.string.yesterday) + " " +
                                                context.getString(R.string.a_las) + " " +
                                                hora
                                    )
                                } else {
                                    // fecha cualquiera
                                    tvStatus.setText(
                                        context.getString(R.string.ultVez) + " " +
                                                fecha + " " +
                                                context.getString(R.string.a_las) + " " +
                                                hora
                                    )
                                }
                            }
                        }
                    }
                } else {
                    // no hay nodo "Estado" → mostrar desconectado
                    iconConnected.setVisibility(View.GONE)
                    iconDisconnected.setVisibility(View.VISIBLE)
                    tvStatus.setText(context.getString(R.string.desconectado)) // "disconnected" → desconectado
                }
            }

            override fun onCancelled(error: DatabaseError) {
                // opcional: log
            }
        })
    }

    fun setNoLeido(id_user: String, type: String) {
        refDatos.child(user!!.uid).child(type).child(id_user).child("noVisto")
            .addListenerForSingleValueEvent(object : ValueEventListener {
                override fun onDataChange(dataVistos: DataSnapshot) {
                    if (dataVistos.exists()) {
                        val noVistos = dataVistos.getValue<Int?>(Int::class.java)

                        refDatos.child(user.uid).child("ChatList").child("msgNoLeidos")
                            .addListenerForSingleValueEvent(object : ValueEventListener {
                                override fun onDataChange(dataLeidos: DataSnapshot) {
                                    if (dataLeidos.exists()) {
                                        val noLeidos = dataLeidos.getValue<Int?>(Int::class.java)
                                        val count = noLeidos!! - noVistos!!
                                        if ((noVistos > 0)) {
                                            dataVistos.getRef().setValue(0)
                                            dataLeidos.getRef().setValue(count)
                                        } else {
                                            dataVistos.getRef().setValue(1)
                                            dataLeidos.getRef().setValue(noLeidos + 1)
                                        }
                                    }
                                }

                                override fun onCancelled(error: DatabaseError) {
                                }
                            })
                    }
                }

                override fun onCancelled(error: DatabaseError) {
                }
            })
    }

    @JvmStatic
    fun Silent(name_user: String?, id_user: String?, type: String?) {
        refDatos.child(user!!.uid).child(type!!).child(id_user!!)
            .addListenerForSingleValueEvent(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    if (dataSnapshot.exists()) {
                        val state =
                            dataSnapshot.child("estado").getValue<String?>(String::class.java)
                        val photo =
                            dataSnapshot.child("wUserPhoto").getValue<String?>(String::class.java)

                        if (photo == Constants.EMPTY) {
                            dataSnapshot.ref.removeValue()
                        } else {
                            if (state == "silent") {
                                dataSnapshot.ref.child("estado").setValue(type)
                            } else {
                                dataSnapshot.ref.child("estado").setValue("silent")
                            }
                        }
                    } else {
                        newChatWith(dataSnapshot, id_user, name_user!!, "silent")
                    }
                }

                override fun onCancelled(error: DatabaseError) {
                }
            })
    }

    fun newChatWith(dataSnapshot: DataSnapshot, id_user: String, name_user: String, state: String) {
        val c = Calendar.getInstance()
        val dateFormat3 = SimpleDateFormat("dd/MM/yyyy HH:mm:ss:SS")
        val newChat = ChatWith(
            "",
            dateFormat3.format(c.getTime()),
            null,
            "",
            id_user,
            name_user,
            Constants.EMPTY,
            state,
            0,
            1
        )

        dataSnapshot.getRef().setValue(newChat)
    }

    @JvmStatic
    fun setBlockUser(
        context: Context,
        nameUser: String?,
        idUser: String?,
        view: View,
        type: String?
    ) {
        if (nameUser == null || idUser == null || type == null) return

        UserMessageUtils.confirm(
            context = context,
            title = "Bloquear",
            message = "¿Desea bloquear a $nameUser?",
            onConfirm = {
                refDatos.child(user!!.uid).child(type).child(idUser)
                    .addListenerForSingleValueEvent(object : ValueEventListener {
                        override fun onDataChange(dataSnapshot: DataSnapshot) {
                            if (dataSnapshot.exists()) {
                                dataSnapshot.ref.child("estado").setValue("bloq")
                            } else {
                                newChatWith(dataSnapshot, idUser, nameUser, "bloq")
                            }
                        }

                        override fun onCancelled(error: DatabaseError) {}
                    })

                UserMessageUtils.showSnack(
                    root = view,
                    message = "Bloqueaste a $nameUser, podrás desbloquearlo cuando desees",
                    duration = Snackbar.LENGTH_INDEFINITE,
                    actionText = "OK",
                    action = {},
                    iconRes = R.drawable.ic_info_24
                )
            }
        )
    }


    @JvmStatic
    fun setUnBlockUser(
        context: Context,
        idUser: String?,
        nameUser: String?,
        view: View,
        type: String?
    ) {
        if (idUser == null || nameUser == null || type == null) return

        UserMessageUtils.confirm(
            context = context,
            title = "Desbloquear",
            message = "¿Desea desbloquear a $nameUser?",
            onConfirm = {
                refDatos.child(user!!.uid).child(type).child(idUser)
                    .addListenerForSingleValueEvent(object : ValueEventListener {
                        override fun onDataChange(dataSnapshot: DataSnapshot) {
                            val photo = dataSnapshot.child("wUserPhoto").getValue(String::class.java)
                            if (photo == Constants.EMPTY) {
                                dataSnapshot.ref.removeValue()
                            } else {
                                dataSnapshot.ref.child("estado").setValue(type)
                            }
                        }

                        override fun onCancelled(error: DatabaseError) {}
                    })

                UserMessageUtils.showSnack(
                    root = view,
                    message = "Desbloqueaste a $nameUser",
                    duration = Snackbar.LENGTH_SHORT,
                    iconRes = R.drawable.ic_info_24
                )
            }
        )
    }



    fun bindBlockStatus(
        user_id: String,
        profile_bloc: ImageView
    ) { // bindBlockStatus = vincular estado de bloqueo

        refDatos.child(user!!.uid).child(Constants.CHATWITH).child(user_id).child("estado")
            .addValueEventListener(object : ValueEventListener {
                override fun onDataChange(dataSnapshot: DataSnapshot) {
                    if (dataSnapshot.getValue<String?>(String::class.java) == "bloq") {
                        profile_bloc.visibility = View.VISIBLE
                    } else {
                        profile_bloc.visibility = View.GONE
                    }
                }

                override fun onCancelled(error: DatabaseError) {
                }
            })
    }

    @JvmField
    var latitude: Double = 0.0
    @JvmField
    var longitude: Double = 0.0
    @JvmStatic
    fun updateLocationUI(mLastLocation: Location?) {
        if (mLastLocation == null) return

        latitude = mLastLocation.latitude
        longitude = mLastLocation.longitude

        refCuentas.child(user!!.uid).addListenerForSingleValueEvent(object : ValueEventListener {
            override fun onDataChange(dataSnapshot: DataSnapshot) {
                if (dataSnapshot.exists()) {
                    FirebaseRefs.refCuentas.child(user.uid).child("latitud").setValue(latitude)
                    FirebaseRefs.refCuentas.child(user.uid).child("longitud").setValue(
                        longitude
                    )
                }
            }

            override fun onCancelled(error: DatabaseError) {}
        })
    }
}
